flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -269,6 +267,17 @@ private MemorySegment requestMemorySegmentFromGlobal() throws IOException {
 			if (segment != null) {
 				numberOfRequestedMemorySegments++;
 				return segment;
+			} else if (isBlocking) {
+				// if the future is completed before the callback is registered,
+				// the request thread will wait 2s before polling an available
+				// segment from the global pool, which is not a big problem.
```

Comment:
Currently, we rely on the wait to release the sync lock, so operations like buffer recycle can acquire the log. If we replace the wait with future.get, we need to consider how to release the sync lock properly.

Reply:
Would that work?
```
	@Nullable
	private MemorySegment requestMemorySegment(boolean isBlocking) throws InterruptedException, IOException, ExecutionException {
		if (!isBlocking) {
			return requestMemorySegment();
		}
		else {
			MemorySegment segment = null;
			while (segment == null) {
				segment = requestMemorySegment();
				if (segment == null) {
					isAvailable().get();
				}
			}
			return segment;
		}
	}

	@Nullable
	private MemorySegment requestMemorySegment() throws IOException {
		MemorySegment segment = null;
		synchronized (availableMemorySegments) {
			returnExcessMemorySegments();

			if (availableMemorySegments.isEmpty()) {
				segment = requestMemorySegmentFromGlobal();
			}
			if (segment == null) {
				segment = availableMemorySegments.poll();
			}
			if (isUnavailable()) {
				availabilityHelper.resetUnavailable();
			}
		}
		return segment;
	}
```
One downside of this is that with the above version, we are calling `returnExcessMemorySegments()` and `availabilityHelper.resetUnavailable();` multiple times per one blocking call - but only if the buffer is not presetn, so probably performance overhead of the extra checks do not matter.

Reply:
I think it should work. The advantage of this implementation is that it can reduce useless wakeup avoid 2 second waiting if the future is completed immediately. 
Maybe we need do tow more things: one is to wrap the ```ExecutionException``` as ```IOException``` or ```InterruptedException``` to avoid the change of method signature (including the caller); the other is to complete the available future when buffer pool is destroyed to unblocking the main thread.

Reply:
> The advantage of this implementation is that it can reduce useless wakeup and avoid 2 second waiting if the future is completed immediately.

I think more importantly, it cleans the code a bit. The previous `if (Thread.holdsLock...)` check inside the runnable action attached to `isAvailable()` looked quite strange, which you correctly spotted, as it required long multi line explanation in the comment.
```
// if the future is completed before the callback is registered,
// the request thread will wait 2s before polling an available
// segment from the global pool, which is not a big problem
```
Also rest of the code is relaying on `isAvailable().get()`, it makes sense to use the same pattern here, instead of `lock.wait()`.

> Maybe we need do tow more things: one is to wrap the ExecutionException as IOException or InterruptedException to avoid the change of method signature (including the caller); the other is to complete the available future when buffer pool is destroyed to unblocking the main thread.

Makes sense, I intentinoally skipped the exceptions but good that you caught the destroying case - I missed that.